Wamd header extractor

noahmacaskill opened the pull request:
Extracted all metadata from the wamd chunk.

atruskie commented on src/MetadataUtility/Audio/Wamd.cs:
```diff
@@ -0,0 +1,203 @@
+        {
+            OffsetDateTime? date = null;
+
+            try
+            {
+                date = OffsetDateTimePattern.CreateWithInvariantCulture("yyyy'-'MM'-'dd' 'HH':'mm':'sso<m>").Parse(value).Value;
+            }
+            catch (UnparsableValueException)
+            {
+                return null;
+            }
```
A couple of notes here:

- we probably want EMU to crash if the string value is not-empty and we can't parse the date; this scenario would be a date format we don't understand and we want to halt, throw an exception, and hopefully get a user that reports the bug so we can improve emu.
- NodaTime specifically returns an type that can be checked for failure, so we can avoid exception-based logic, why is this wrapper even necesary?

On the assumption that the answer to the above is because it fails when  `value` is null or empty, I'd suggest checking that first, and returning null in those cases before attempting to do the parsing.

noahmacaskill replied:
Fixed this up but wanted to clarify. As of my most recent changes it should crash with an unrecognized date format. I could instead use some Fin logic to log the error and keep going?

atruskie replied:
Unrecognized format? In the Wamd chunk?

Yeah an exception I think.

In a filename, we'd not crash, but that's because we're attempting parse dozens of possible formats for which there may not even be a date included! 

In wamd the behavior is clearly defined by a doc so any behavior outside of that is exceptional.

atruskie commented on src/MetadataUtility/Audio/Wamd.cs:
```diff
@@ -0,0 +1,203 @@
+        /// <param name="wamdData">The wamd object.</param>
+        public static void SetLocation(string value, Wamd wamdData)
+        {
+            string[] locationInfo = value.Split(",", StringSplitOptions.RemoveEmptyEntries);
```
hmm why is empty entry removal needed here?

atruskie replied:
Yeah I'm betting it is because altitude is often empty. We should explicitly attempt to parse altitude in case it is ever not empty. I'm assuming it would be populated when a GPS attachment is added to a SM4 for example.

noahmacaskill replied:
This one's odd, this is how location is presented according to the documentation:

`WGS84,nn.nnnnn,N,mmm.mmmmm,W[,alt]`

I think WGS84 is just the coordinate system being used? In all the examples I've seen that space is empty, meaning location starts with a leading comma, this is what I was filtering out. In case that first space is ever populated with something I'm no longer filtering, just ignoring whatever is in that first space.

I've also added support for altitude.

atruskie replied:
It might be worth noting something about that first space being a coordinate system that is WGS84 and if it is ever not empty we should add code to handle it. This is a good place for a `Debug.assert`, or an exception.

atruskie commented on src/MetadataUtility/Audio/Wamd.cs:
```diff
@@ -0,0 +1,203 @@
+            // Links metadata set functions to their corresponding subchunk ID
+            Dictionary<int, Action<string>> setters = new Dictionary<int, Action<string>>
+            {
+                { 1, value => wamdData.Name = value },
+                { 2, value => wamdData.SerialNumber = value },
+                { 3, value => wamdData.Firmware = value },
+                { 5, value => wamdData.StartDate = DateParser(value) },
+                { 18, value => wamdData.MicrophoneType = value.Split(",") },
+                { 19, value => wamdData.MicrophoneSensitivity = value.Split(",") },
+                { 20, value => SetLocation(value, wamdData) },
+                { 21, value => wamdData.Temperature = value },
```
Some points:  

- make an explicit effort to name those magic numbers (and also format them as hex over decimal) as they are in the wamd header file - makes it much easier to cross reference them
- You've got a mix between mutative and pure functions doing the parsing? Why? I'd probably lean always towards Pure to isolate effects
- This dictionary can be statically defined once at the class level. Defining it in this method means it is always defined on every invocation of the method.
- And why the partial implementation?

noahmacaskill replied:
By partial implementation did you just mean I'm not accounting for every value in their metadata doc? Pulled what I thought were all the important metadata that I saw in the files I have. Didn't think anything else in there correlated to metadata fields that we have. The only ones that could be interesting but weren't in any of the files I've seen are humidity or light? Could add metadata fields for those, or implement others if there are any you think are of value?

atruskie replied:
You took a practical approach. I prefer myself to create fully functional primitives.

One eventual goal of EMU is to be able to strip any metadata out of any file and then allow said files to be converted without loss of information. That might be a good reason to fully deserialize the chunk.

But if that doesn't align with current goals I'm happy if you leave it until later.

atruskie commented on src/MetadataUtility/Audio/Wamd.cs:
```diff
@@ -0,0 +1,203 @@
+                int start = wamdOffset;
+                int end = wamdOffset + (int)length;
+
+                value = Encoding.ASCII.GetString(wamdSpan[start..end]);
```
HMMM some fields could be UTF-8 https://github.com/QutEcoacoustics/emu/blob/b132d18c1f188c461446c490cc6ece40fadfee7d/docs/WildlifeAcousticsMetadataHeader/wa_meta.h#L97-L99

Not sure what to do about that...

atruskie replied:
It looks like none of the current definitions use that provision, but maybe leave a todo to future editors that utf-8 may need to be selectively used for decoding the value.

atruskie commented on src/MetadataUtility/Metadata/WildlifeAcoustics/WamdExtractor.cs:
```diff
@@ -0,0 +1,78 @@
+
+        public ValueTask<bool> CanProcessAsync(TargetInformation information)
+        {
+            var result = information.IsPcmWaveFile() && information.HasVersion1WamdChunk();
```
Theoretically a .wac file could also have a wamd chunk. Also would be true for WA new proprietary file format (which I forget the name of). Leave a note for future editiors?

// TODO: add support for .wac (or other WA) files?

atruskie commented on src/MetadataUtility/Models/Microphone.cs:
```diff
@@ -39,5 +39,10 @@ public record Microphone
         /// Gets the name of the microphone (style depends on vendor).
         /// </summary>
         public string ChannelName { get; init; }
+
+        /// <summary>
+        /// Gets the sensitivity of the microphone.
+        /// </summary>
+        public string Sensitivity { get; init; }
```
we should make a note of the correct SI unit for this and make sure we're reporting the units correctly. Also document the unit which this value is in.

atruskie commented on src/MetadataUtility/Models/Sensor.cs:
```diff
@@ -63,5 +63,10 @@ public record Sensor
         /// configuration file.
         /// </summary>
         public string[] Configuration { get; init; }
+
+        /// <summary>
+        /// Gets the temperature from the sensor.
+        /// </summary>
+        public string Temperature { get; init; }
```
Again, document units, and make sure we're reporting units a SI units (degrees celsius)

atruskie commented on test/MetadataUtility.Tests/Audio/WamdTests.cs:
```diff
@@ -0,0 +1,30 @@
+        {
+        }
+
+        [Theory]
+        [ClassData(typeof(FixtureHelper.FixtureData))]
+        public void HasVersion1WamdChunkTest(FixtureModel model)
+        {
+            bool hasWamd = Wamd.HasVersion1WamdChunk(model.ToTargetInformation(this.RealFileSystem).FileStream).IfFail(false);
+
+            hasWamd.Should().Be(model.CanProcess.Contains("WamdExtractor"));
+        }
```
OK this looks incomplete. 

For the low-level tests, pick one file (fixture) and write tests to make sure the low level chunk parser is working

Don't: use our fixture dataset
Don't: use the fixture model
Don't: test CanProcess or Process - they are functions of the extractor (not the subject of this test file)

atruskie reviewed (changes requested):
Looks great. More work to be done, but it's just filling in the gaps. No major architectural work needed.

noahmacaskill commented:
Got started on this today. For documenting units, the units of temperature and microphone sensitivity are celsius and decibels relative to full scale (dBFS) respectively. The given temperature values are marked with a C but microphone sensitivity is unmarked so I tacked dBFS on the end for those values. Also marked the unit in a comment above the field, could incorporate this into the name instead (e.g. TemperatureCelsius)?

Not too familiar with microphone sensitivity but doing some reading on it, it seems that dB is an SI unit, but dBFS is not. Going to read more about it tomorrow but maybe I'll convert into mV/Pa?

atruskie commented:
Hmmm. 

I think dBFS is fine. The base unit is SI and the FS is just indicating it is a ratio anchored to an digital amplitude of 1.0. Actually as I understand it all dB measurements are a ratio and without an anchor they can only be interpreted relatively.

I actually don't think we could convert to mV/Pa? We'd need to know detailed information about the electronics of the sensor?

This is beyond my knowledge. I think we should just document the unit in the docstring and if we find another sensor that reports the units differently then we'll deal with the conversion then.

As for naming, I really want to avoid adding an unit to the names in serialized structures. If the need arises I was thinking about encoding the value as an object/string to make it clear. But again, I think just basic impl first with docs and we can adjust later as the need arises.

noahmacaskill commented:
Had a go at addressing everything, there were a couple things I'm unsure on.

I implemented an Either monad for ambiguous/unambiguous dates. I added a LocalStartDate field to the recoding model which I'm setting the case where only a local date is given, is that what you had in mind?

noahmacaskill commented on test/MetadataUtility.Tests/Audio/WamdTests.cs:
```diff
@@ -0,0 +1,55 @@
+            hasWamd.Should().Be(model.CanProcess.Contains("WamdExtractor"));
+        }
+
+        [Fact]
+        public void ExtractMetadataTest()
+        {
+            Stream stream = this.RealFileSystem.File.Open(Helpers.FixturesRoot + "/WA_SM4BAT/2.2.1_Normal/S4U09523_20210621_205706.wav", FileMode.Open, FileAccess.Read, FileShare.Read);
+
+            var tryWamdData = Wamd.ExtractMetadata(stream);
+
+            Assert.True(tryWamdData.IsSucc);
+
+            Wamd wamdData = (Wamd)tryWamdData;
+
+            ((OffsetDateTime)wamdData.StartDate).Should().Be(OffsetDateTimePattern.CreateWithInvariantCulture("G").Parse("2021-06-21T20:57:06-03:00").Value);
+            wamdData.Name.Should().Be("SM4BAT-FS");
+            wamdData.SerialNumber.Should().Be("S4U09523");
+            wamdData.Firmware.Should().Be("2.2.1");
+            wamdData.Temperature.Should().Be("24.25C");
+            wamdData.MicrophoneType.Should().AllBe("U2");
+            wamdData.MicrophoneSensitivity.Should().AllBe("13.0dBFS");
+            wamdData.Latitude.Should().Be(45.7835);
+            wamdData.Longitude.Should().Be(-64.23352);
+        }
+    }
```
Made another wamd test. Not sure if this was what you had in mind, let me know if I'm off base here.

atruskie replied:
looks good. the above HasVersion1WamdChunkTest really doesn't need to be a theory. Rather than test every file, just pick two files to test true and false for `HasVersion1WamdChunk`

atruskie commented:
> Had a go at addressing everything, there were a couple things I'm unsure on.
> 
> I implemented an Either monad for ambiguous/unambiguous dates. I added a LocalStartDate field to the recoding model which I'm setting the case where only a local date is given, is that what you had in mind?

I think I was going to hardline it originally and make it so the date would only show up if it was unambiguous or if it --utc-offset was passed in and used on the CLI

Showing local date as well is probably more useful to more people, though I'm worried it will encourage the same bad habits people have been doing since the start.

Leave it for now. I might remove it later. We'll see.

atruskie commented:
This is still marked as a draft. Mark it as ready and I'll merge
